lua/auto-hlsearch.lua

```diff
@@ -1,23 +1,38 @@
+local M = {}
+
 local defaults = {
   remap_keys = { "/", "?", "*", "#", "n", "N" },
 }
 
--- Remap provided keys in order to use :AutoHlsearch command
+-- Remap provided keys in order to use activate() function
 -- while keeping the user's keymap configuration
 local function remap_keys(keys)
   local function set(lhs, keymap)
-    local cmd = ":AutoHlsearch<CR>"
-    if keymap and keymap.rhs then
+    if keymap then
       local opts = {
-        expr = keymap.expr,
+        expr = true,
         noremap = keymap.noremap,
         nowait = keymap.nowait,
         script = keymap.script,
         silent = keymap.silent,
       }
-      vim.api.nvim_set_keymap("n", lhs, string.format("%s%s", cmd, keymap.rhs), opts)
+      -- We need to consider the remmaping when use expr options
+      -- For lua function
+      if keymap.callback then
+        vim.keymap.set("n", lhs, function () M.activate() return keymap.callback() end, opts)
+
+      elseif keymap.expr == 1 and keymap.rhs then
+        -- For vimscript function
+        vim.keymap.set("n", lhs, function () M.activate() return vim.api.nvim_eval(keymap.rhs) end , opts)
+
+      -- For vimscript function, not use expr options
+      elseif keymap.rhs then
+          vim.keymap.set("n", lhs, function () M.activate() return keymap.rhs end, opts)
```

Comment:
Interesting.. if `expr=true`, the return from lua callback would get evaluated? Didn't know about that. I tested it and it works, but is that documented somewhere in the nvim help? I didn't find anything about it.

Reply:
vim.keymap.set()'s help is writen like "Accepts options accepted by the {opts} parameter in nvim_set_keymap()"
then nvim_set_keymap()'s help is writen like the below
When called from Lua, also accepts a "callback" key that takes a Lua function to call when the mapping is executed.
When "expr" is true, "replace_keycodes" (boolean) can be used to replace keycodes in the resulting string (see nvim_replace_termcodes()), and a Lua callback returning `nil` is equivalent to returning an empty string.

Reply:
Oh, I see, thank you!